### ISSUE TITLE
Right value to put in hostname and domainname when FQDN is a fourth-level-domain

### ISSUE THREAD

**Issue by frugan-it** [docker-mailserver/docker-mailserver#2423]
# Subject

I would like some feedback concerning a use case

## Description

I am a little confused as to what is the right value to put in `hostname` and `domainname` in `docker-compose.yml`, when FQDN is a **fourth-level-domain** (eg. `sub4.sub3.domain.tld`).
I searched the issues a bit, and read these https://github.com/docker-mailserver/docker-mailserver/pull/2245#issuecomment-942919301, https://github.com/docker-mailserver/docker-mailserver/pull/2280, but I don't find my specific case.

I am currently using this setting in `docker-compose.yml`:

```yml
version: '3.9'

services:
  mailserver:
    image: docker.io/mailserver/docker-mailserver:10.4.0
    container_name: mailserver
    hostname: sub4.sub3.domain.tld
    domainname: domain.tld
...
```
but it does not seem correct to me, in fact it causes some small problems:

1. `/etc/mailname` contains `sub3.domain.tld`, and not `sub4.sub3.domain.tld`
2. UNCHECKED mail notifications by Amavis are sent to postmaster@sub3.domain.tld, and not postmaster@sub4.sub3.domain.tld
3. PostSRSd rewrites `Return-Path: <SRS0=xxxx=yy=otherdomain.tld=user@sub3.domain.tld>`, and not `Return-Path: <SRS0=xxxx=yy=otherdomain.tld=user@sub4.sub3.domain.tld>`

However, I think points 2) and 3) depend on point 1), in fact about Amavis:

```bash
$ grep -R postmaster /etc/amavis/conf.d/
conf.d/20-debian_defaults:$virus_admin = "postmaster\@$mydomain"; # due to D_DISCARD default
```
and

```bash
$ cat /etc/amavis/conf.d/05-domain_id 
...
chomp($mydomain = `head -n 1 /etc/mailname`);
...
```
And about PostSRSd:

https://github.com/roehling/postsrsd/blob/master/init/postsrsd-systemd-launcher.in#L4
```
SRS_DOMAIN="`postconf -h mydomain || true`"
```
and
```
$ postconf -h mydomain
sub3.domain.tld
```

**Comment by casperklein** [comment]
Did you try this:

```
services:
  mailserver:
    hostname: sub4
    domainname: sub3.domain.tld
```

**Comment by polarathene** [comment]
**TL;DR:** Try just using `hostname`, in most cases you should have no need for `domainname` being configured:

```
hostname: sub4.sub3.domain.tld
```

---

````
hostname: sub4.sub3.domain.tld
domainname: domain.tld
```

Overlapping like that is not advised anywhere AFAIK? IIRC that's handled poorly.

`domainname` is primarily treated as an NIS domain I believe. You should be able to ignore it entirely and place the entire FQDN in `hostname` unless you have an issue caused by that, in which case share that with us :)

```
hostname: sub4.sub3.domain.tld
# You don't need `domainname`:
#domainname: domain.tld
```

In many cases you should expect that the 1st DNS label (sub4) is treated as the short hostname (`hostname -s`) and the remainder (`sub3.domain.tld`) your domain (output of commands: `hostname -d` / `dnsdomainname`). I also made [a useful reference table here](https://github.com/docker-mailserver/docker-mailserver/pull/2245#issuecomment-942919301) if you need to troubleshoot further.

One of those shows how to handle a multi-label short hostname if you absolutely require that (unlikely), it's called "multi-label split + /etc/hosts fixed" and would look like this:

```
hostname: sub4.sub3
domainname: domain.tld
# Also requires fixing your /etc/hosts file in the container.
# Note this didn't investigate compatibility with `docker-mailserver` and may fail with software like postfix trying to make sense of it too.
```

---

> And about PostSRSd

We have logic that configures this AFAIK:

https://github.com/docker-mailserver/docker-mailserver/blob/358df6ada2e405652fa7c33c4c46dbd9f414d7f0/target/scripts/postsrsd-wrapper.sh#L8-L17

First our `_obtain_hostname_and_domainname` method is called which will derive the internal `HOSTNAME` and `DOMAINNAME` vars, often from the CLI/yaml options for `hostname`/`domainname`, and there is also a DMS ENV we support which has priority `SRS_DOMAINNAME`.

One of those values is then chosen to overwrite a config file for postsrsd, and that should be the one that is used.

**Comment by frugan-it** [comment]
Hi there, thanks for the replies!

Without detracting from the contribution of @casperklein, I tried to put:

```
hostname: sub4.sub3.domain.tld
```
that it would seem to me the most intuitive setting..
And of course to restart the container, but nothing has changed.. :confused:

```
$ hostname
sub4.sub3.domain.tld

$ cat /etc/mailname
sub3.domain.tld

$ postconf -h mydomain
sub3.domain.tld
```

I knew of the existence of vars `SRS_DOMAINNAME`, and also `OVERRIDE_HOSTNAME`, but I was hoping I didn't have to use them, and that the system would detect the correct settings.

At this point I begin to think, as far as I understand, that these values are correct, what do you say?
Is it right for Amavis to send notifications to `postmaster@sub3.domain.tld` ?
And that PostSRSd rewrites `Return-Path: <SRS0=mrqb=TC=otherdomain.tld=user@sub3.domain.tld>` ?

**Comment by polarathene** [comment]
> And of course to restart the container, but nothing has changed..

Just to clarify. AFAIK, you can have mail for `example.test` and send email to `hello@example.test`, where the mail server itself is configured for `mail.example.test`. In this case is `sub4` meant to be equivalent to `mail` for you? Do you actually want that as part of the email address? Otherwise your MX record will handle that part.

> At this point I begin to think, as far as I understand, that these values are correct, what do you say?
> Is it right for Amavis to send notifications to `postmaster@sub3.domain.tld` ?
> And that PostSRSd rewrites `Return-Path: <SRS0=mrqb=TC=otherdomain.tld=user@sub3.domain.tld>` ?

I ironically don't run a mail server myself, I've only contributed and help maintain at present. That looks correct and as intended to the best of my knowledge though, unless you weren't happy with that. @casperklein can probably confirm :)

> I knew of the existence of vars `SRS_DOMAINNAME`, and also `OVERRIDE_HOSTNAME`, but I was hoping I didn't have to use them, and that the system would detect the correct settings.

- `SRS_DOMAINNAME` was only contributed AFAIK for some niche cases where the user wanted to use a different value than `docker-mailserver` would use elsewhere for the domain name.
- `OVERRIDE_HOSTNAME` was introduced as a workaround, and IIRC was due more to a lack of understanding of how `hostname`/`domainname` are handled by the container OS and our own scripts/packages. If someone finds that works for them, there's probably a better way to resolve it and once I've got the time to verify, it'll be dropped in a future release.

Technically the correct settings are being used. `mydomain` is the FQDN minus the short hostname or 1st DNS label as far as I know, in this case that's `sub4` being dropped. If you want `mydomain` to behave differently, you could configure it explicitly to do so, have you found that `docker-mailserver` is behaving differently than an alternative mail setup with Postfix?

**Comment by frugan-it** [comment]
Thanks for your detailed answer @polarathene, and for the time you all dedicate to this fantastic project! :pray:

I just think I had wrong expectations, due to my lack of skills in this regard, but as you suggest, the result is consistent with what was set.

I am left with only one doubt: why, while setting `DOMAINNAME` as `domain.tld`, my file `/etc/mailname` contains `sub3.domain.tld`?
Shouldn't the [_setup_mailname](https://github.com/docker-mailserver/docker-mailserver/blob/ec8b99335ea3d1d9754217cf4826582caf0c0bec/target/scripts/startup/setup-stack.sh#L135) function take care of this?

Alternatively i was thinking of using `user-patches.sh` to insert `domain.tld` in `/etc/mailname` instead of `sub3.domain.tld`.
Do you think that's the right approach?

Unlike what I wrote before, I would like to configure as `postmaster@domain.tld` system mails that are not affected by the value of `POSTMASTER_ADDRESS` (like Amavis, PostSRSd, etc.), but which apparently derive from the value of `/etc/mailname`.

**Comment by polarathene** [comment]
> I am left with only one doubt: why, while setting `DOMAINNAME` as `domain.tld`, my file `/etc/mailname` contains `sub3.domain.tld`?
> Shouldn't the [_setup_mailname](https://github.com/docker-mailserver/docker-mailserver/blob/ec8b99335ea3d1d9754217cf4826582caf0c0bec/target/scripts/startup/setup-stack.sh#L135) function take care of this?

It is working correctly. It will receive `DOMAINNAME` (internal var we manage) from:

https://github.com/docker-mailserver/docker-mailserver/blob/ec8b99335ea3d1d9754217cf4826582caf0c0bec/target/scripts/helper-functions.sh#L197-L202

https://github.com/docker-mailserver/docker-mailserver/blob/ec8b99335ea3d1d9754217cf4826582caf0c0bec/target/scripts/helper-functions.sh#L219-L220

https://github.com/docker-mailserver/docker-mailserver/blob/ec8b99335ea3d1d9754217cf4826582caf0c0bec/target/scripts/helper-functions.sh#L234-L242

https://github.com/docker-mailserver/docker-mailserver/blob/ec8b99335ea3d1d9754217cf4826582caf0c0bec/target/scripts/helper-functions.sh#L246-L248

So basically, we rely on `hostname -d` usually, at least for anything our scripts do.

Local / system mail is usually sent via a user without any address part after an `@`, services like Postfix IIRC are configured to append something like `mydomain` value to generate an actual e-mail address for the sender in the mail being sent. My memory of that is a bit foggy.

You shouldn't set `DOMAINNAME` yourself, only `domainname:` (docker-compose) or `--domainname` (Docker CLI), although I advise just using `hostname:`/`--hostname` instead and letting the domain name be extracted.. just note that if you want a short hostname that is more than one label (`sub4.sub3`) that's against specs IIRC and may not work well (_see the referenced table from earlier, which required modifying the containers `/etc/hosts` file to use the correct multi-label short hostname_).

You can also have Postfix be configured for aliases (system ones, not user virtual aliases) via `/tmp/docker-mailserver/postfix-aliases.cf`, which should allow specifying which accounts should map to a specific address. Another maintainer may be able to better advise here about this (@casperklein ?) :)

I'm not sure about changing `/etc/mailname` either.

**Comment by frugan-it** [comment]
Thank you so much @polarathene for your patiences! :pray:

Now everything is clearer: earlier I thought the `DOMAINNAME` variable was derived from the `domainame` variable set via docker-compose.. :trollface:

However I set `domain.tld` in `/etc/mailname` via `user-patches.sh`, and now Amavis notifications (eg. UNCHECKED emails) are sent correctly to `postmaster@domain.tld` (and no longer to `postmaster@sub3.domain.tld`) .
Surely, as you write, you can achieve the same result with Postfix aliases, or maybe even adding this in `/etc/amavis/conf.d/50-user`:

```
$virus_admin = "postmaster\@domain.tld";
```

but I preferred `mailname` route for now..

**Comment by polarathene** [comment]
You're very welcome :)

Glad this appears to be resolved now? Should the issue be closed? Other maintainers will still chime in if they have any other additional input/advice.

**Comment by frugan-it** [comment]
Yes thanks, even if I haven't solved it very elegantly, the problem is still solved..
